### PR TITLE
Publishing init images for k8s

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?

Adds publishing of release images to github packages, dockerhub, GCR, and ECR.

### Motivation

This way lib injection can be done by the admission controller.

### Additional Notes

Draft for now since:

1. Repos need to be created/renamed on dockerhub and ECR.
2. Need to create a multiarch image.

### CODE REVIEW

File: .github/workflows/release-latest.yml [hunk 1]
```diff
@@ -28,6 +28,22 @@ jobs:
       - run: |
           git tag v${{ fromJson(steps.pkg.outputs.json).version }}
           git push origin v${{ fromJson(steps.pkg.outputs.json).version }}
+      - name: Set up Docker Buildx
```
Comment (rochdev) [inline .github/workflows/release-latest.yml:31]: This is not part of npm publish, so it should be in a separate job. Also, it should be moved to a separate action so that it can be used in the 2.x release as well.

File: .github/workflows/release-latest.yml [hunk 2]
```diff
@@ -28,6 +28,22 @@ jobs:
...
+        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # 2.0.0
+        with:
+          install: true
+          config-inline: |
+            [worker.oci]
+              max-parallelism = 1
```
Comment (rochdev) [inline .github/workflows/release-latest.yml:37]: Why limit parallelism here?

File: .github/workflows/release-latest.yml [hunk 3]
```diff
@@ -28,6 +28,22 @@ jobs:
...
+      - name: Build injection image and push to github packages
+        run: |
+          node scripts/set-version-for-init-image.js ${{ fromJson(steps.pkg.outputs.json).version }}
+          npm pack dd-trace@${{ fromJson(steps.pkg.outputs.json).version }}
```
Comment (rochdev) [inline .github/workflows/release-latest.yml:41]: Should this have been done in the npm publish process and stored as an artifact?

File: .github/workflows/release-latest.yml [hunk 4]
```diff
@@ -28,6 +28,22 @@ jobs:
...
+          cp dd-trace-${{ fromJson(steps.pkg.outputs.json).version }}.tgz k8s/auto-inst/dd-trace.tgz
```
Comment (rochdev) [inline .github/workflows/release-latest.yml:42]: Why not pack there directly or move it?

File: .github/workflows/release-latest.yml [hunk 5]
```diff
@@ -28,6 +28,22 @@ jobs:
...
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # 2.0.0
+        with:
+          install: true
+          config-inline: |
+            [worker.oci]
+              max-parallelism = 1
+      - name: Build injection image and push to github packages
+        run: |
+          node scripts/set-version-for-init-image.js ${{ fromJson(steps.pkg.outputs.json).version }}
```
Comment (rochdev) [inline .github/workflows/release-latest.yml:40]: Why is this needed?

File: .github/workflows/test-k8s-lib-injection.yaml [hunk 6]
```diff
@@ -39,8 +39,10 @@ jobs:
         run: |
           npm pack
           cp dd-trace-*.tgz k8s/auto-inst/dd-trace.tgz
-          docker build k8s/auto-inst --build-arg npm_pkg=./dd-trace.tgz -t ghcr.io/datadog/dd-trace-js/dd-trace-js-init:${GITHUB_SHA}
-          docker push ghcr.io/datadog/dd-trace-js/dd-trace-js-init:${GITHUB_SHA}
+          docker buildx create --name lib-injection
+          docker buildx use lib-injection
+          docker run --rm --privileged multiarch/qemu-user-static --reset -p yes # https://stackoverflow.com/questions/72167570/docker-buildx-nodejs-fail
+          docker buildx build --platform=linux/amd64,linux/arm/v7,linux/arm64/v8 --build-arg npm_pkg=./dd-trace.tgz -t ghcr.io/datadog/dd-trace-js/dd-lib-js-init:${GITHUB_SHA} --push k8s/auto-inst
```
Comment (rochdev) [inline .github/workflows/test-k8s-lib-injection.yaml:45]: Should this test use the exact same action as the release process to ensure that we're testing what we ship?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: .github/workflows/release-latest.yml (hunk 5 above)
Comment (bengl) [reply .github/workflows/release-latest.yml:40]: This changes the version in the `package.json` so that the correct version of the tracer is installed.

I could instead do this as a build arg in docker, which would prevent all of this process. Would you prefer that?

File: .github/workflows/release-latest.yml (hunk 4 above)
Comment (bengl) [reply .github/workflows/release-latest.yml:42]: A move is probably more correct. I wanted to avoid switching directories for simplicity. I'll change it to a move.

File: .github/workflows/release-latest.yml (hunk 3 above)
Comment (bengl) [reply .github/workflows/release-latest.yml:41]: When referencing a package and version, `npm pack` will download the tarball without installing it. (or pull it from cache). That's what we're going for here.

File: .github/workflows/release-latest.yml (hunk 1 above)
Comment (bengl) [reply .github/workflows/release-latest.yml:31]: Tbh this is my workaround for not knowing how to have jobs dependent on one another :cry:. I'll have another look.

File: .github/workflows/release-latest.yml (hunk 2 above)
Comment (bengl) [reply .github/workflows/release-latest.yml:37]: This is copypasta from the Java version of this.

https://github.com/DataDog/dd-trace-java/blob/1e9abac39fbc42906f1cb107cdf64e47074d692c/.github/workflows/test-lib-injection.yaml#L45-L51

File: .github/workflows/test-k8s-lib-injection.yaml (hunk 6 above)
Comment (bengl) [reply .github/workflows/test-k8s-lib-injection.yaml:45]: It does now! The only difference is whether or not the `--push` arg is provided.

Would you rather I encapsulate this in a script so that it's less likely to diverge?

Comment (bengl) [reply .github/workflows/test-k8s-lib-injection.yaml:45]: Err, no, now I see what you're saying. Let me go ahead and do that.

File: .github/workflows/release-latest.yml (hunk 1 above)
Comment (rochdev) [reply .github/workflows/release-latest.yml:31]: You can use `needs: ['publish']` similar to the `docs` job.

File: .github/workflows/release-latest.yml (hunk 5 above)
Comment (rochdev) [reply .github/workflows/release-latest.yml:40]: I'm not sure what that would look like so hard to say. What would be the benefit of the current approach compared to the build arg?

Comment (bengl) [reply .github/workflows/release-latest.yml:40]: As it turns out this is way simpler than I had previously thought. In my next rev of this (gonna push in a few min) this won't need any build args, nor will it need this script. It will just pick up whatever tarball is around and use that.

**Review by rochdev — APPROVED** [review]
